compiler/rustc_hir_typeck/src/pat.rs

```diff
@@ -83,6 +83,9 @@ struct PatInfo<'tcx, 'a> {
     binding_mode: BindingMode,
     top_info: TopInfo<'tcx>,
     decl_origin: Option<DeclOrigin<'a>>,
+
+    parent_kind: Option<PatKind<'tcx>>,
+    current_kind: Option<PatKind<'tcx>>,
```

Comment:
Since you only care whether the parent exists and not its kind, you could store a single boolean here, names `has_parent` or `is_top_level` or something like that.

Reply:
I used a counter `current_depth` to distinct three states:

* Initial state: 0
* The first call of `check_pat`: 1
* Second and subsequent calls of `check_pat`: >1

If `current_depth` is less than or equal to 1, the pattern is top level.